File: src/services/api/notificationsApi.js
import { apiRequest, ensureAccessToken, getStoredAuth } from './client'

const hasNotificationSession = () => {
  const auth = getStoredAuth()
  return Boolean(auth?.accessToken && auth?.user && !auth.user.isAnonymous)
}

export const getNotifications = async (limit = 30) => {
  if (!hasNotificationSession()) {
    return []
  }

  const token = await ensureAccessToken()
  if (!token) {
    return []
  }

  try {
    return await apiRequest(`/notifications?limit=${limit}`, { token })
  } catch (error) {
    if (error.status === 401) {
      return []
    }
    throw error
  }
}

export const markNotificationRead = async (id) => {
  const token = await ensureAccessToken()
  if (!token) {
    return null
  }

  return apiRequest(`/notifications/${encodeURIComponent(id)}/read`, {
    method: 'PATCH',
    token,
  })
}

export const registerPushToken = async (token, permission = 'granted') => {
  const accessToken = await ensureAccessToken()
  if (!accessToken) {
    return null
  }

  return apiRequest('/notifications/push-token', {
    method: 'POST',
    token: accessToken,
    body: {
      token,
      permission,
      platform: 'web',
    },
  })
}

export const unregisterPushToken = async (token) => {
  const accessToken = await ensureAccessToken()
  if (!accessToken) {
    return null
  }

  return apiRequest('/notifications/push-token', {
    method: 'DELETE',
    token: accessToken,
    body: { token },
  })
}

File: app_movile/lib/features/notifications/data/notifications_api.dart
import '../../../core/api/api_client.dart';
import 'app_notification.dart';

class NotificationsApi {
  NotificationsApi(this._client);

  final ApiClient _client;

  Future<List<AppNotification>> getNotifications({int limit = 40}) async {
    final payload = await _client.request(
      '/notifications?limit=$limit',
      token: _client.accessToken,
    );

    if (payload is! List) {
      return const [];
    }

    return payload
        .whereType<Map<String, dynamic>>()
        .map(AppNotification.fromJson)
        .where((item) => item.id.isNotEmpty)
        .toList(growable: false);
  }

  Future<void> markRead(String id) async {
    await _client.request(
      '/notifications/${Uri.encodeComponent(id)}/read',
      method: 'PATCH',
      token: _client.accessToken,
    );
  }

  Future<void> registerPushToken({
    required String token,
    required String platform,
    String permission = 'granted',
  }) async {
    await _client.request(
      '/notifications/push-token',
      method: 'POST',
      token: _client.accessToken,
      body: {
        'token': token,
        'platform': platform,
        'permission': permission,
      },
    );
  }

  Future<void> unregisterPushToken(String token) async {
    await _client.request(
      '/notifications/push-token',
      method: 'DELETE',
      token: _client.accessToken,
      body: {'token': token},
    );
  }
}
